src/frontend/qt_sdl/main.cpp

```diff
@@ -1399,16 +1403,52 @@ QString MainWindow::loadErrorStr(int error)
     }
 }
 
+std::string extractROM(char* zipName, std::string zipDir){
+    //Open the ZIP archive
+    int err = 0;
+    zip *z = zip_open(zipName, 0, &err);
+    
+	struct zip_stat st;
+    zip_stat_init(&st);
+    zip_stat_index(z, 0, 0, &st); //Get information about the file at index 0
+	
+    //Allocate memory for its uncompressed contents
+    char *contents = new char[st.size];
+	
+    //Read the compressed file
+    zip_file *f = zip_fopen_index(z, 0, 0); //Open file at index 0
+	zip_fread(f, contents, st.size);
+    zip_fclose(f);
+
+    zip_close(z);
+	
+	//Write the file (binary mode)
+    std::ofstream(zipDir + "/" + st.name, std::ofstream::binary).write(contents, st.size);
```

Comment:
throwing it onto the disk and then reading it again is not optimal, especially if it's just directly next to the zip file and not some temp folder.

We currently don't have any interface from the core to load a rom from a blob. We already discussed this internally, it's probably the best to get @Arisotura involved as well

Reply:
yup. I have ideas for this (providing a ROMSource interface that the core will use to access the ROM, or whatever name is good for you; then frontends will implement classes derived from that interface, to implement loading from a file, from a ZIP, etc...)

archives also pose the problem that we do need to let the user select a file from within the archive

Reply:
> providing a ROMSource interface that the core will use to access the ROM

wouldn't this mean you would have to change a lot of the code. most of the file work is done using fseek, fread which take file pointers.

> archives also pose the problem that we do need to let the user select a file from within the archive

shouldn't be too hard. i could just read all the files inside the zip, if there's one, launch it, or if there's more, open a dialog box. 

> and also, where do save files and all go?

what about putting them in a folder with the same name of the zip file. for example, Mario Kart DS.nds inside DS_ROMS.zip would be placed in DS_ROMS/Mario Kart DS.sav.